simple_api/object/actions.py

```diff
@@ -18,19 +19,30 @@ def set_name(self, name):
 
     def set_parent_class(self, cls):
         self.parent_class = cls
-        for field in {**self.parameters, **self.data}.values():
+        for field in self.parameters.values():
+            field.set_parent_class(cls)
+        for field in self.data.values():
             field.set_parent_class(cls)
         if self.return_value is not None:
             self.return_value.set_parent_class(cls)
 
 
 class Action(SetReferencesMixin, ToActionMixin):
-    def __init__(self, parameters=None, data=None, return_value=None, exec_fn=None, permissions=None, **kwargs):
+    def __init__(self, parameters=None, data=None, return_value=None, exec_fn=None, permissions=None, validators=None,
+                 **kwargs):
         self.parameters = parameters or {}
         self.data = data or {}
         self.return_value = return_value
         self.exec_fn = exec_fn
         self.permissions = permissions or ()
+        self.action_validators = validators or ()
+
+        # all field validators are read from argument and restructured into [("parameter", (validator1, validator2...))]
+        self.field_validators = []
+        if parameters is not None:
+            self.field_validators = [(x, parameters[x].validators) for x in parameters
+                                     if parameters[x].validators is not None]
```

Comment:
jak asi tusis, tohle je potencialne problemove...jako vstupni data mame `params` (zhruba odpovida url parametrum z restu) a `data` (zhruba odpovida request body z restu); zde resis jenom parametry - nevim, jestli pro data je to todo, nebo jsi na to zapomnel :)

zbytek je code review, nevim, jak moc jsi zbehly v pythonu, kdyztak se neco naucis: `x = a or b` ulozi do promenne `x` hodnotu `a`, pokud neni `None`, 0, `""`, `False` nebo jeste mozna neco (asi chapes), jinak se ulozi `b`. Tedy tento trik pujde jednoduse pouzit k tomu, abys nemusel mit `if` v tom generatoru - staci proste zaridit, aby se misto `None` kdyztak ulozila prazdna kolekce...na danem miste je dalsi comment

proc je to pole dvojic a ne slovnik? jinak pres slovnik iteruj pomoci `for key, value in dict.items()` (misto key a value pojmenuj promenne nejak smysluplne podle toho, co znamenaji) - nikdo nebude zkoumat, co znamena `x[0]` a `x[1]`

Reply:
No já si nebyl jistý jak přesně to s `data` je, už jsem tam přidal že to prochází i validátory tam.

Nejsem si úplně jistý jak nahradit v tom list (nebo teď dict comprehension) or hodnotu místo if, chápu jak to funguje, ale když tam dám cokoliv tak to zbytečně přidá do list/dict a budu pak přes to zbytečně iterovat když teď ne. Nebo je tam něco co mi uniká? Přeci jen jsem aktivně na Python skoro vůbec od maturity nešáhnul...

Reply:
Mělo by plně fungovat i na data